File: apps/android/app/src/main/java/chat/rogi/rogichat/core/media/MediaClient.kt
package chat.rogi.rogichat.core.media

import java.net.URI
import kotlinx.coroutines.delay
import kotlinx.serialization.json.*

/** Executed by the existing authenticated gateway under its original immutable permit.
 * It must guard admission, HTTP and commit, cancel on revocation, reject redirects, bound JSON
 * to 1 MiB, enforce expectedStatus and stream upload as exact application/octet-stream.
 * No API bearer may be forwarded to object storage. */
fun interface MediaTransport { suspend fun perform(request: MediaRequest, scope: MediaScope): String }
class MediaRequest internal constructor(val method: String, val path: String, val expectedStatus: Int,
                                       val jsonBody: String? = null, val upload: MediaFile? = null)
class MediaClient(private val transport: MediaTransport, val scope: MediaScope) {
    private suspend fun request(method: String, path: String, status: Int, body: String? = null, upload: MediaFile? = null): String {
        scope.check(); upload?.validate()
        val result = transport.perform(MediaRequest(method, path, status, body, upload), scope)
        scope.check(); require(result.toByteArray(Charsets.UTF_8).size <= 1_048_576)
        return result
    }
    private fun json(body: String) = body
    suspend fun reserve(file: MediaFile): MediaReceipt {
        require(file.kind != MediaKind.AVATAR || scope.roomId == null)
        file.validate()
        return MediaReceipt.decode(request("POST", "media/upload-intents", 201,
            json(MediaIntent(file.kind, file.contentType, file.byteLength, if (file.kind == MediaKind.AVATAR) null else scope.roomId).json())))
            .also { check(it.status == MediaStatus.reserved) }
    }
    suspend fun upload(assetId: String, file: MediaFile): MediaReceipt {
        mediaId(assetId); file.validate()
        return MediaReceipt.decode(request("POST", "media/upload-intents/$assetId/content", 202, upload = file))
            .also { check(it.assetId == assetId && it.status == MediaStatus.processing) }
    }
    suspend fun status(assetId: String): MediaReceipt = MediaReceipt.decode(request("GET", "media/upload-intents/${mediaId(assetId)}", 200))
        .also { check(it.assetId == assetId) }
    /** Bounded wait; a timeout leaves a recoverable asset ID, never invents success or reuploads. */
    suspend fun awaitReady(assetId: String, attempts: Int = 60): MediaReceipt {
        require(attempts in 1..120)
        repeat(attempts) { val receipt = status(assetId)
            when (receipt.status) { MediaStatus.ready -> return receipt
                MediaStatus.deleting, MediaStatus.deleted -> throw MediaUnavailable()
                else -> delay(2000) }
        }
        throw MediaProcessingTimeout()
    }
    suspend fun access(assetId: String, context: MediaAccess): MediaLease {
        val room = when (context) { is MediaAccess.Message -> context.roomId; is MediaAccess.Avatar -> context.roomId
            is MediaAccess.Sticker -> context.roomId; is MediaAccess.Preview -> null }
        require(room == null || room == scope.roomId)
        val started = System.nanoTime()
        val result = Json.parseToJsonElement(request("POST", "media/assets/${mediaId(assetId)}/access", 200, json(context.json()))).jsonObject
        check(result.getValue("expiresIn").jsonPrimitive.int == 60)
        val url = result.getValue("url").jsonPrimitive.content
        val uri = URI(url); require(uri.scheme == "https" && !uri.host.isNullOrEmpty() && uri.userInfo == null && uri.fragment == null)
        return MediaLease(url, started + 55_000_000_000L, context.variant)
    }
    suspend fun renewAccess(assetId: String, context: MediaAccess, replacing: MediaLease): MediaLease {
        replacing.checkedURL(scope)
        return kotlinx.coroutines.withTimeout(replacing.renewalBudgetMillis()) {
            access(assetId, context).also { it.checkedURL(scope) }
        }
    }
    suspend fun stickers(after: String? = null): MediaStickerPage {
        val room = mediaId(requireNotNull(scope.roomId))
        val path = "rooms/$room/stickers" + (after?.let { "?after=${mediaId(it)}" } ?: "")
        val o = Json.parseToJsonElement(request("GET", path, 200)).jsonObject
        val items = o.getValue("items").jsonArray.map { val s = it.jsonObject
            MediaSticker(s.getValue("id").jsonPrimitive.content, s.getValue("label").jsonPrimitive.content, s.getValue("assetId").jsonPrimitive.content) }
        val next = o.getValue("nextCursor").jsonPrimitive.contentOrNull?.let(::mediaId)
        return MediaStickerPage(items, next)
    }
    /** PATCH only the avatar field so concurrent nickname/birthday edits are not overwritten. */
    suspend fun updateAvatar(ready: MediaReceipt?) {
        require(scope.roomId == null)
        require(ready == null || ready.status == MediaStatus.ready)
        val raw = request("PATCH", "me/profile", 200, json(buildJsonObject { put("avatarAssetId", ready?.assetId?.let(::JsonPrimitive) ?: JsonNull) }.toString()))
        val result = Json.parseToJsonElement(raw).jsonObject
        mediaId(result.getValue("id").jsonPrimitive.content)
        val avatar = result.getValue("avatar")
        check(if (ready == null) avatar == JsonNull else avatar.jsonObject.getValue("assetId").jsonPrimitive.content == ready.assetId)
    }
}

File: apps/ios/Sources/Core/Media/MediaClient.swift
import Foundation

// NativeAPIClient owner mounts this request through the existing original credential/permit.
// Adapter checks admission/HTTP/commit, cancels on invalidation, rejects redirects, bounds JSON
// to 1 MiB, enforces expectedStatus and streams upload.url as exact application/octet-stream.
struct MediaRequest: Sendable {
    let method: String
    let path: String
    let expectedStatus: Int
    let jsonBody: Data?
    let upload: MediaFile?
    fileprivate init(_ method: String, _ path: String, _ status: Int, body: Data? = nil, upload: MediaFile? = nil) {
        self.method = method; self.path = path; self.expectedStatus = status; self.jsonBody = body; self.upload = upload
    }
}
protocol MediaTransport: Sendable { func perform(_ request: MediaRequest, scope: any MediaScope) async throws -> Data }
struct MediaClient: Sendable {
    let transport: any MediaTransport
    let scope: any MediaScope
    var apiBaseURL: URL? = nil
    private func request(_ request: MediaRequest) async throws -> Data {
        try scope.check(); try Task.checkCancellation(); try request.upload?.validate()
        let result = try await transport.perform(request, scope: scope)
        try scope.check(); try Task.checkCancellation()
        guard result.count <= 1_048_576 else { throw MediaError.invalid }; return result
    }
    func reserve(_ file: MediaFile) async throws -> MediaReceipt {
        guard file.kind != .avatar || scope.roomID == nil else { throw MediaError.invalid }
        let intent = try MediaIntent(file: file, roomID: file.kind == .avatar ? nil : scope.roomID)
        let data = try await request(MediaRequest("POST", "media/upload-intents", 201, body: JSONEncoder().encode(intent)))
        let receipt = try JSONDecoder().decode(MediaReceipt.self, from: data).validated()
        guard receipt.status == .reserved else { throw MediaError.invalid }; return receipt
    }
    func upload(_ assetID: String, file: MediaFile) async throws -> MediaReceipt {
        let path = "media/upload-intents/\(try mediaID(assetID))/content"
        let data = try await request(MediaRequest("POST", path, 202, upload: file))
        let receipt = try JSONDecoder().decode(MediaReceipt.self, from: data).validated(assetID: assetID)
        guard receipt.status == .processing else { throw MediaError.invalid }; return receipt
    }
    func status(_ assetID: String) async throws -> MediaReceipt {
        let data = try await request(MediaRequest("GET", "media/upload-intents/\(try mediaID(assetID))", 200))
        return try JSONDecoder().decode(MediaReceipt.self, from: data).validated(assetID: assetID)
    }
    func awaitReady(_ assetID: String, attempts: Int = 60) async throws -> MediaReceipt {
        guard (1...120).contains(attempts) else { throw MediaError.invalid }
        for _ in 0..<attempts {
            let receipt = try await status(assetID)
            switch receipt.status {
            case .ready: return receipt
            case .deleting, .deleted: throw MediaError.unavailable
            default: try await Task.sleep(for: .seconds(2))
            }
        }
        throw MediaError.processing
    }
    func access(_ assetID: String, context: MediaAccess) async throws -> MediaLease {
        guard context.roomID == nil || context.roomID == scope.roomID else { throw MediaError.invalid }
        let started = ContinuousClock.now
        let data = try await request(MediaRequest("POST", "media/assets/\(try mediaID(assetID))/access", 200, body: context.body()))
        struct Receipt: Decodable { let url: URL; let expiresIn: Int }
        let receipt = try JSONDecoder().decode(Receipt.self, from: data)
        guard receipt.expiresIn == 60 else { throw MediaError.invalid }
        return try MediaLease(url: receipt.url, started: started, variant: context.variant)
    }
    func renewAccess(_ assetID: String, context: MediaAccess, replacing: MediaLease) async throws -> MediaLease {
        _ = try replacing.checkedURL(scope: scope)
        let budget = replacing.renewalBudget()
        return try await withThrowingTaskGroup(of: MediaLease.self) { group in
            group.addTask { let lease = try await access(assetID, context: context); _ = try lease.checkedURL(scope: scope); return lease }
            group.addTask { try await Task.sleep(for: budget); throw MediaError.expired }
            defer { group.cancelAll() }
            guard let lease = try await group.next() else { throw MediaError.expired }; return lease
        }
    }
    func providerAvatar(actorID: String? = nil) async throws -> MediaLease {
        let path: String
        if let actorID {
            guard let room = scope.roomID else { throw MediaError.invalid }
            path = "rooms/\(try mediaID(room))/actors/\(try mediaID(actorID))/provider-avatar/access"
        } else {
            guard scope.roomID == nil else { throw MediaError.invalid }
            path = "me/provider-avatar/access"
        }
        let started = ContinuousClock.now
        let data = try await request(MediaRequest("POST", path, 200))
        struct Receipt: Decodable { let url: URL; let expiresIn: Int }
        let receipt = try JSONDecoder().decode(Receipt.self, from: data)
        guard receipt.expiresIn == 60 else { throw MediaError.invalid }
        try validateProviderAvatarURL(receipt.url, apiBaseURL: apiBaseURL)
        return try MediaLease(url: receipt.url, started: started, variant: .image)
    }
    func stickers(after: String? = nil) async throws -> MediaStickerPage {
        guard let room = scope.roomID else { throw MediaError.invalid }
        var path = "rooms/\(try mediaID(room))/stickers"
        if let after { path += "?after=\(try mediaID(after))" }
        let data = try await request(MediaRequest("GET", path, 200))
        let page = try JSONDecoder().decode(MediaStickerPage.self, from: data)
        for item in page.items { _ = try mediaID(item.id); _ = try mediaID(item.assetId) }
        if let next = page.nextCursor { _ = try mediaID(next) }
        return page
    }
    func updateAvatar(_ ready: MediaReceipt?) async throws {
        guard scope.roomID == nil else { throw MediaError.invalid }
        if let ready { _ = try ready.validated(); guard ready.status == .ready else { throw MediaError.invalid } }
        struct Body: Encodable {
            let avatarAssetId: String?
            enum CodingKeys: CodingKey { case avatarAssetId }
            func encode(to encoder: any Encoder) throws {
                var c = encoder.container(keyedBy: CodingKeys.self)
                if let avatarAssetId { try c.encode(avatarAssetId, forKey: .avatarAssetId) } else { try c.encodeNil(forKey: .avatarAssetId) }
            }
        }
        let data = try await request(MediaRequest("PATCH", "me/profile", 200, body: JSONEncoder().encode(Body(avatarAssetId: ready?.assetId))))
        struct Acknowledgement: Decodable {
            struct Avatar: Decodable { let assetId: String }
            let id: String
            let avatar: Avatar?
            enum CodingKeys: CodingKey { case id, avatar }
            init(from decoder: any Decoder) throws {
                let c = try decoder.container(keyedBy: CodingKeys.self)
                id = try c.decode(String.self, forKey: .id)
                avatar = try c.decode(Avatar?.self, forKey: .avatar)
            }
        }
        let result = try JSONDecoder().decode(Acknowledgement.self, from: data)
        _ = try mediaID(result.id)
        guard result.avatar?.assetId == ready?.assetId else { throw MediaError.invalid }
    }
}

func validateProviderAvatarURL(_ url: URL, apiBaseURL: URL?) throws {
    guard let apiBaseURL, let base = URLComponents(url: apiBaseURL, resolvingAgainstBaseURL: false),
          let value = URLComponents(url: url, resolvingAgainstBaseURL: false),
          base.scheme == "https", base.host != nil, base.user == nil, base.password == nil,
          base.percentEncodedPath == "/v1/", base.query == nil, base.fragment == nil,
          value.scheme == "https", value.host == base.host, value.port == base.port,
          value.user == nil, value.password == nil, value.fragment == nil,
          value.percentEncodedPath == "/v1/profile-images", let query = value.percentEncodedQuery,
          query.range(of: "^ticket=[A-Za-z0-9_-]{64,1024}$", options: .regularExpression) != nil else { throw MediaError.invalid }
}
